#
# Incorporate some 10-STABLE speed-ups of temp files and a Utah change
# to pw_user to not make so many group file writes.
#
# cd /usr/src
# sudo patch -p0 < patchfile
#

diff -Nur lib/libutil.orig/gr_util.c lib/libutil/gr_util.c
--- lib/libutil.orig/gr_util.c	2015-08-13 16:38:22.562033943 -0600
+++ lib/libutil/gr_util.c	2016-12-19 15:02:48.150513000 -0700
@@ -25,7 +25,7 @@
  */
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: releng/10.2/lib/libutil/gr_util.c 285205 2015-07-06 13:01:01Z garga $");
+__FBSDID("$FreeBSD: stable/10/lib/libutil/gr_util.c 310173 2016-12-16 20:10:55Z asomers $");
 
 #include <sys/param.h>
 #include <sys/errno.h>
@@ -141,7 +141,7 @@
 		errno = ENAMETOOLONG;
 		return (-1);
 	}
-	if ((tfd = mkostemp(tempname, O_SYNC)) == -1)
+	if ((tfd = mkostemp(tempname, 0)) == -1)
 		return (-1);
 	if (mfd != -1) {
 		while ((nr = read(mfd, buf, sizeof(buf))) > 0)
diff -Nur lib/libutil.orig/pw_util.c lib/libutil/pw_util.c
--- lib/libutil.orig/pw_util.c	2015-08-13 16:38:22.522033209 -0600
+++ lib/libutil/pw_util.c	2016-12-19 15:02:18.455217000 -0700
@@ -39,7 +39,7 @@
 static const char sccsid[] = "@(#)pw_util.c	8.3 (Berkeley) 4/2/94";
 #endif
 static const char rcsid[] =
-  "$FreeBSD: releng/10.2/lib/libutil/pw_util.c 285205 2015-07-06 13:01:01Z garga $";
+  "$FreeBSD: stable/10/lib/libutil/pw_util.c 310173 2016-12-16 20:10:55Z asomers $";
 #endif /* not lint */
 
 /*
@@ -226,7 +226,7 @@
 		errno = ENAMETOOLONG;
 		return (-1);
 	}
-	if ((tfd = mkostemp(tempname, O_SYNC)) == -1)
+	if ((tfd = mkostemp(tempname, 0)) == -1)
 		return (-1);
 	if (mfd != -1) {
 		while ((nr = read(mfd, buf, sizeof(buf))) > 0)
diff -Nur usr.sbin/pw.orig/Makefile usr.sbin/pw/Makefile
--- usr.sbin/pw.orig/Makefile	2015-08-13 16:36:55.272035347 -0600
+++ usr.sbin/pw/Makefile	2016-11-22 12:52:29.668849000 -0700
@@ -8,6 +8,8 @@
 	grupd.c pwupd.c fileupd.c psdate.c \
 	bitmap.c cpdir.c rm_r.c
 
+#CFLAGS+=-DOLDCODE_COMPAT
+
 WARNS?=	3
 
 DPADD=	${LIBCRYPT} ${LIBUTIL} ${LIBSBUF}
diff -Nur usr.sbin/pw.orig/grupd.c usr.sbin/pw/grupd.c
--- usr.sbin/pw.orig/grupd.c	2015-08-13 16:36:55.382035309 -0600
+++ usr.sbin/pw/grupd.c	2016-12-19 15:06:45.469362000 -0700
@@ -26,16 +26,15 @@
 
 #ifndef lint
 static const char rcsid[] =
-  "$FreeBSD: releng/10.2/usr.sbin/pw/grupd.c 285092 2015-07-03 14:22:44Z bapt $";
+  "$FreeBSD: stable/10/usr.sbin/pw/grupd.c 310173 2016-12-16 20:10:55Z asomers $";
 #endif /* not lint */
 
+#include <err.h>
 #include <grp.h>
 #include <libutil.h>
-#include <err.h>
 #include <stdio.h>
 #include <stdlib.h>
-#include <string.h>
-#include <sys/param.h>
+#include <unistd.h>
 
 #include "pwupd.h"
 
@@ -75,8 +74,11 @@
 	}
 	if (gr_copy(pfd, tfd, gr, old_gr) == -1) {
 		gr_fini();
+		close(tfd);
 		err(1, "gr_copy()");
 	}
+	fsync(tfd);
+	close(tfd);
 	if (gr_mkdb() == -1) {
 		gr_fini();
 		err(1, "gr_mkdb()");
diff -Nur usr.sbin/pw.orig/pw_nis.c usr.sbin/pw/pw_nis.c
--- usr.sbin/pw.orig/pw_nis.c	2015-08-13 16:36:55.272035347 -0600
+++ usr.sbin/pw/pw_nis.c	2016-12-19 15:05:26.747052000 -0700
@@ -26,13 +26,15 @@
 
 #ifndef lint
 static const char rcsid[] =
-  "$FreeBSD: releng/10.2/usr.sbin/pw/pw_nis.c 285092 2015-07-03 14:22:44Z bapt $";
+  "$FreeBSD: stable/10/usr.sbin/pw/pw_nis.c 310173 2016-12-16 20:10:55Z asomers $";
 #endif /* not lint */
 
 #include <sys/types.h>
+
 #include <err.h>
 #include <pwd.h>
 #include <libutil.h>
+#include <unistd.h>
 
 #include "pw.h"
 
@@ -43,6 +45,7 @@
 	struct passwd *pw = NULL;
 	struct passwd *old_pw = NULL;
 
+	printf("===> %s\n", path);
 	if (pwd != NULL)
 		pw = pw_dup(pwd);
 
@@ -61,8 +64,11 @@
 	}
 	if (pw_copy(pfd, tfd, pw, old_pw) == -1) {
 		pw_fini();
+		close(tfd);
 		err(1, "pw_copy()");
 	}
+	fsync(tfd);
+	close(tfd);
 	if (chmod(pw_tempname(), 0644) == -1)
 		err(1, "chmod()");
 	if (rename(pw_tempname(), path) == -1)
diff -Nur usr.sbin/pw.orig/pw_user.c usr.sbin/pw/pw_user.c
--- usr.sbin/pw.orig/pw_user.c	2015-08-13 16:36:55.262049646 -0600
+++ usr.sbin/pw/pw_user.c	2016-11-22 12:52:29.670114000 -0700
@@ -636,6 +636,7 @@
 
 	if (mode == M_ADD || getarg(args, 'G') != NULL) {
 		int i, j;
+#ifdef OLDCODE_COMPAT
 		/* First remove the user from all group */
 		SETGRENT();
 		while ((grp = GETGRENT()) != NULL) {
@@ -668,8 +669,55 @@
 			chggrent(cnf->groups[i], grp);
 			free(grp);
 		}
-	}
+#else
+		/*
+		 * N.B. every chggrent() call causes a copy of the entire group
+		 * file, so minimize the number of calls we make!
+		 */
+		SETGRENT();
+		while ((grp = GETGRENT()) != NULL) {
+			int oingroup, ningroup;
+			char group[MAXLOGNAME];
 
+			strlcpy(group, grp->gr_name, MAXLOGNAME);
+
+			/* Are we supposed to be in this group? */
+			ningroup = 0;
+			for (i = 0; cnf->groups[i] != NULL; i++) {
+				if (strcmp(cnf->groups[i], group) == 0) {
+					ningroup = 1;
+					break;
+				}
+			}
+
+			/* Are we already in this group? */
+			oingroup = -1;
+			for (i = 0; grp->gr_mem && grp->gr_mem[i] != NULL; i++) {
+				if (strcmp(grp->gr_mem[i], pwd->pw_name) == 0) {
+					oingroup = i;
+					break;
+				}
+			}
+
+			/* If we found the user and they should not be here, remove */
+			if (oingroup >= 0 && ningroup == 0) {
+				for (j = oingroup; grp->gr_mem[j] != NULL; j++)
+					grp->gr_mem[j] = grp->gr_mem[j+1];
+				chggrent(group, grp);
+			}
+
+			/* If we did not find the user and they should be here, add */
+			else if (oingroup < 0 && ningroup != 0) {
+				grp = gr_add(grp, pwd->pw_name);
+				if (grp != NULL) {
+					chggrent(group, grp);
+					free(grp);
+				}
+			}
+		}
+		ENDGRENT();
+#endif
+	}
 
 	/* go get a current version of pwd */
 	pwd = GETPWNAM(name);
diff -Nur usr.sbin/pw.orig/pwupd.c usr.sbin/pw/pwupd.c
--- usr.sbin/pw.orig/pwupd.c	2015-08-13 16:36:55.282035386 -0600
+++ usr.sbin/pw/pwupd.c	2016-12-19 15:10:21.840546000 -0700
@@ -26,21 +26,19 @@
 
 #ifndef lint
 static const char rcsid[] =
-  "$FreeBSD: releng/10.2/usr.sbin/pw/pwupd.c 285092 2015-07-03 14:22:44Z bapt $";
+  "$FreeBSD: stable/10/usr.sbin/pw/pwupd.c 310173 2016-12-16 20:10:55Z asomers $";
 #endif /* not lint */
 
+#include <sys/wait.h>
+
+#include <err.h>
+#include <errno.h>
+#include <pwd.h>
+#include <libutil.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
-#include <pwd.h>
-#include <libutil.h>
-#include <errno.h>
-#include <err.h>
-#include <sys/types.h>
-#include <sys/stat.h>
-#include <sys/param.h>
-#include <sys/wait.h>
 
 #include "pwupd.h"
 
@@ -113,8 +111,11 @@
 	}
 	if (pw_copy(pfd, tfd, pw, old_pw) == -1) {
 		pw_fini();
+		close(tfd);
 		err(1, "pw_copy()");
 	}
+	fsync(tfd);
+	close(tfd);
 	/*
 	 * in case of deletion of a user, the whole database
 	 * needs to be regenerated
